### PR TITLE
adding instance method initialization

### PR DESCRIPTION
**Pull request by walterddr**
Currently ScalarTransformFunctionWrapper doesn't actually take advantage of Literal arguments passed in. 

Creating a framework (very basic) to showcase how to invoke initialization to reduce overhead during each scalarFunction invoke.
This is using a very basic example of string normalize function to reduce the overhead of each match of ENUM extract, not much but it will come really handy in other more heavy use cases like regex Pattern compilation (which is per row)

TODO
- [ ] create a interface for the init method (or making it accept variable-length arguments like `Object... `?)
- [ ] sanitize the APIs so that there's less if checkers

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapper.java [hunk 1]
```diff
@@ -113,6 +114,12 @@ public void init(List<TransformFunction> arguments, Map<String, DataSource> data
       }
     }
     _nonLiteralValues = new Object[_numNonLiteralArguments][];
+
+    // if FunctionInvoker contains an instance function. initialize it.
+    Object instance = _functionInvoker.getInstance();
+    if (instance != null && instance instanceof InitializableScalarFunction) {
```
Comment (richardstartin) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapper.java:120]: `instance instanceof InitializableScalarFunction` will fail (i.e. yield false) if the instance is null, so the null check is superfluous.

Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapper.java:120]: did not know that. good catch

**Comment by richardstartin** [comment]
I don't know how you feel about this (I could see the case for keeping it separate), but to prove this is really useful it might be nice to include a couple of applications in this PR, e.g. regex and JsonPath. Generally I like to go from the specific to the abstract, and having a few specific examples handled nicely by the abstraction gives me more confidence in its fitness for purpose.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 2]
```diff
@@ -328,14 +329,27 @@ public static String normalize(String input) {
...
+    }
+
+    /**
+     * see Normalizer#normalize(String, Form)
+     * @param input
+     * @param form
+     * @return transforms string with the specified normalization form
+     */
+    @ScalarFunction
+    String normalize(String input, String form) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:350]: There are other callers to `FunctionInvoker.invoke()`, and we need to make sure all of them call `init()` first

**Comment by walterddr** [comment]
this doesn't work. it need to support both function with the exact same signature but one expects literal string to passed in during query optimization, another expects a normal string during ingestion

**Pull request closed without merging**